host/src/filesystem.rs

```diff
@@ -330,21 +331,52 @@ impl wasi_filesystem::WasiFilesystem for WasiCtx {
         &mut self,
         fd: wasi_filesystem::Descriptor,
     ) -> HostResult<wasi_filesystem::DirEntryStream, wasi_filesystem::Errno> {
-        todo!()
+        let iterator = self
+            .table()
+            .get_dir(fd)
+            .map_err(convert)?
+            .readdir(ReaddirCursor::from(0))
+            .await
+            .map_err(convert)?;
+
+        self.table_mut()
+            .push(Box::new(Mutex::new(iterator)))
+            .map_err(convert)
     }
 
     async fn read_dir_entry(
         &mut self,
         stream: wasi_filesystem::DirEntryStream,
     ) -> HostResult<Option<wasi_filesystem::DirEntry>, wasi_filesystem::Errno> {
-        todo!()
+        let entity = self
+            .table()
+            .get::<Mutex<ReaddirIterator>>(stream)
+            .map_err(convert)?
+            .lock()
+            .unwrap()
+            .next()
+            .transpose()
+            .map_err(convert)?;
+
+        Ok(entity.map(|e| wasi_filesystem::DirEntry {
+            ino: Some(e.inode),
+            type_: e.filetype.into(),
+            name: e.name,
+        }))
     }
 
     async fn close_dir_entry_stream(
         &mut self,
-        fd: wasi_filesystem::DirEntryStream,
+        stream: wasi_filesystem::DirEntryStream,
     ) -> anyhow::Result<()> {
-        todo!()
+        // Verify we're deleting an entry of the expected type:
+        self.table()
+            .get::<Mutex<ReaddirIterator>>(stream)
+            .map_err(convert)?;
+
+        self.table_mut().delete(stream);
```

Comment:
I think we're always going to want to verify the type when deleting a table entry; would it make sense to add a type parameter to the delete function and have it do the verification?